scripts/us_hud/income/process.py

```diff
@@ -64,68 +61,136 @@ def get_url(year):
+    workbook = python_calamine.CalamineWorkbook.from_filelike(
+        file)  # type: ignore[arg-type]
+    rows = iter(workbook.get_sheet_by_index(0).to_python())
+    headers = list(map(str, next(rows)))  # Get headers from the first row
+    for row in rows:
+        yield dict(zip(headers, row))
 
-  Args:
-    df: Input dataframe (will be modified).
-    person: Number of people in household.
-  '''
+
+def compute_150(df, person):
+    '''Compute 150th percentile income in-place.'''
     df[f'l150_{person}'] = df.apply(
         lambda x: round(x[f'l80_{person}'] / 80 * 150), axis=1)
 
 
-def process(year, matches, output_dir):
-    '''Generate cleaned CSV.
-
-  Args:
-    year: Input year.
-    matches: Map of fips dcid -> city dcid.
-    output_dir: Directory to write cleaned CSV.
-  '''
+def process(year, matches, input_folder):
+    '''Generate cleaned data and accumulate it in output_data.'''
     url = get_url(year)
-    try:
-        df = pd.read_excel(url)
-    except:
-        print(f'No file found for {url}.')
-        return
+
+    if year == 2023 or year == 2024:
+        try:
+            filename = f"Section8-FY{year}.xlsx"
+            # Read the Excel file and process the generator output
+            with open(os.path.join(input_folder, filename), 'rb') as f:
+                rows = iter_excel_calamine(f)
+                data = list(rows)  # Convert the generator to a list of rows
+            df = pd.DataFrame(data)  # Now create the DataFrame
+        except Exception as e:
+            logging.fatal(f'Error in the process method : {year}: {url} {e}.')
+            return
+    else:
+        # For other years, download via URL
+        try:
+            filename = f"Section8-FY{year}.xls"
+            df = pd.read_excel(os.path.join(input_folder, filename))
+        except Exception as e:
+            logging.fatal(f'Error in the process method : {url} {e}.')
+            return
+
+    # Process the DataFrame (common code for all years)
     if 'fips2010' in df:
         df = df.rename(columns={'fips2010': 'fips'})
 
-    # Filter to 80th percentile income stats for each household size.
+    # Filter to 80th percentile income stats for each household size
     df = df.loc[:, [
         'fips', 'l80_1', 'l80_2', 'l80_3', 'l80_4', 'l80_5', 'l80_6', 'l80_7',
         'l80_8'
     ]]
 
-    df['fips'] = df.apply(lambda x: 'dcs:geoId/' + str(x['fips']).zfill(10),
+    # Format FIPS codes
+    df['fips'] = df.apply(lambda x: 'dcs:geoId/' + str(x['fips']).zfill(5),
                           axis=1)
     df['fips'] = df.apply(lambda x: x['fips'][:-5]
                           if x['fips'][-5:] == '99999' else x['fips'],
                           axis=1)
+
+    # Compute 150th percentile for each household size
     for i in range(1, 9):
         compute_150(df, i)
-    df['year'] = [year for i in range(len(df))]
 
-    # Add stats for matching dcids.
+    # Add year column
+    df['year'] = [year for _ in range(len(df))]
```

Comment:
Can this be also set as
`df['year'] = year`

Reply:
Done and checked